com.unity.ml-agents/Editor/MLAgentsSettingsBuildProvider.cs

```diff
@@ -0,0 +1,74 @@
+using System.Linq;
+using UnityEditor;
+using UnityEditor.Build;
+using UnityEditor.Build.Reporting;
+
+
+namespace Unity.MLAgents.Editor
+{
+    internal class MLAgentsSettingsBuildProvider : IPreprocessBuildWithReport, IPostprocessBuildWithReport
+    {
+        private MLAgentsSettings m_SettingsAddedToPreloadedAssets;
+
+        public int callbackOrder => 0;
+
+        public void OnPreprocessBuild(BuildReport report)
+        {
+            var wasDirty = IsPlayerSettingsDirty();
+            m_SettingsAddedToPreloadedAssets = null;
+
+            var preloadedAssets = PlayerSettings.GetPreloadedAssets().ToList();
+            if (!preloadedAssets.Contains(MLAgentsManager.Settings))
+            {
+                m_SettingsAddedToPreloadedAssets = MLAgentsManager.Settings;
+                preloadedAssets.Add(m_SettingsAddedToPreloadedAssets);
+                PlayerSettings.SetPreloadedAssets(preloadedAssets.ToArray());
+            }
+
+            if (!wasDirty)
+                ClearPlayerSettingsDirtyFlag();
+        }
+
+        public void OnPostprocessBuild(BuildReport report)
+        {
+            if (m_SettingsAddedToPreloadedAssets == null)
+                return;
+
+            var wasDirty = IsPlayerSettingsDirty();
+
+            var preloadedAssets = PlayerSettings.GetPreloadedAssets().ToList();
+            if (preloadedAssets.Contains(m_SettingsAddedToPreloadedAssets))
+            {
+                preloadedAssets.Remove(m_SettingsAddedToPreloadedAssets);
+                PlayerSettings.SetPreloadedAssets(preloadedAssets.ToArray());
+            }
+
+            m_SettingsAddedToPreloadedAssets = null;
+
+            if (!wasDirty)
+                ClearPlayerSettingsDirtyFlag();
+        }
+
+
+        private static bool IsPlayerSettingsDirty()
+        {
+#if UNITY_2019_OR_NEWER
+            var settings = Resources.FindObjectsOfTypeAll<PlayerSettings>();
+            if (settings != null && settings.Length > 0)
+                return EditorUtility.IsDirty(settings[0]);
+            return false;
+#else
```

Comment:
EditorUtility.IsDirty() is only supported in 2019+.

This part of code is borrowed from Input Package Settings and I'm not completely sure what would go wrong if not reset the dirty flag here. I tried with 2018.4 and didn't see any obvious bug.

This is going into April release so maybe I can target v2-staging and get rid of these #if#else.

Reply:
Yeah, shouldn't be a problem - 2018.4 is going away soon.